htscodecs/rANS_static32x16pr_avx512.c

```diff
@@ -564,8 +640,45 @@ unsigned char *rans_compress_O1_32x16_avx512(unsigned char *in,
         last2 = c2;
         iN1 = _mm512_sub_epi32(iN1, _mm512_set1_epi32(1));
         iN2 = _mm512_sub_epi32(iN2, _mm512_set1_epi32(1));
-        c1 = _mm512_i32gather_epi32(iN1, in, 1);
-        c2 = _mm512_i32gather_epi32(iN2, in, 1);
+
+        // Code below is equivalent to this:
+        // c1 = _mm512_i32gather_epi32(iN1, in, 1);
+        // c2 = _mm512_i32gather_epi32(iN2, in, 1);
+
+        // Better when we have a power of 2
+        if (next_batch >= 0) {
+            if (--next_batch < 0 && iN[0] > BATCH) {
+                int i, j;
+                uint8_t c[32][BATCH];
+                iN[0] += BATCH;
+                for (j = 0; j < 32; j++) {
+                    iN[j] -= BATCH;
+                    memcpy(c[j], &in[iN[j]-BATCH], BATCH);
+                }
+                // transpose matrix
+                for (j = 0; j < 32; j++) {
+                    for (i = 0; i < BATCH; i+=16) {
+                        for (int z = 0; z < 16; z++)
+                            t32[i+z][j] = c[j][i+z];
+                    }
+                }
+                next_batch = BATCH-1;
+            }
+            if (next_batch >= 0) {
+              __m128i c1_ = _mm_load_si128((__m128i *)&t32[next_batch][0]);
+              __m128i c2_ = _mm_load_si128((__m128i *)&t32[next_batch][16]);
+              c1 = _mm512_cvtepu8_epi32(c1_);
+              c2 = _mm512_cvtepu8_epi32(c2_);
+            }
+        }
+        if (next_batch < 0) {
+            c1 = _mm512_i32gather_epi32(iN1, in, 1);
```

Comment:
You didn't make these `_mm512_i32gather_epi32x()`.  Was that deliberate?

Reply:
Yes.  Note the last argument is size, which is 1 here.  The _mm512_i32gather_epi32x as written is explicitly only doing 32-bit ints.  If I recall I did experiment with an alternative implementation doing size of 1 too, but it didn't help.